src/mesh/voxel_cloud.rs

```diff
@@ -12,6 +14,1082 @@ use crate::plane::Plane;
 
 use super::{primitive, tools, Face, Mesh, NormalStrategy};
 
+/// Row in the marching cubes lookup table.
+///
+/// Fields are u8 so that more rows fit in an x86 cache line (64 bytes).
+/// With the current layout optimization, a row is 16 bytes, meaning 4 of them fit.
+struct Row {
+    /// Length of a valid data slice.
+    len: u8,
+    /// Row data.
+    data: [(u32, u32, u32); 5],
+}
+
+// FIXME: explain why this is 512 bits rather than expected 488 bits
+static_assertions::assert_eq_size!(Row, [u8; 64]);
```

Comment:
For this struct to be 16 bytes, you need the data to be `u8` also (exactly as the doc comment says), and later do `u32::from` to convert to a wider value when using. Using `u32` in the table itself defeats some the purpose of this optimization - the way it is currently, one row occupies the whole cache line. Not bad, but not good either.

> // FIXME: explain why this is 512 bits rather than expected 488 bits

These are usually expressed in bytes, not bits. And the reason is alignment. Read up:

- https://doc.rust-lang.org/nomicon/repr-rust.html
- https://doc.rust-lang.org/nomicon/other-reprs.html

Reply:
I intentionally used `u32` to avoid casting for each voxel. Is casting (or generally your version with `u8`) more efficient in this case?

Reply:
If nothing anything else, the table (and therefore our binary) is smaller with u8. I measured locally with both u32 and u8 and it does not seem to be much faster with u8 - most likely, because we don't access the table sequentially.

It for sure won't be slower, if that is what you ask.